File: src/leetcode/bit-manipulation.ts
/**
 * 位操作相关算法实现
 * Bit Manipulation related algorithms
 */

/**
 * 136. 只出现一次的数字
 * Single Number
 */
export function singleNumber(nums: number[]): number {
    let result = 0;
    for (const num of nums) {
        result ^= num;
    }
    return result;
}

/**
 * 137. 只出现一次的数字 II
 * Single Number II
 */
export function singleNumberII(nums: number[]): number {
    let ones = 0;
    let twos = 0;
    
    for (const num of nums) {
        ones = (ones ^ num) & ~twos;
        twos = (twos ^ num) & ~ones;
    }
    
    return ones;
}

/**
 * 260. 只出现一次的数字 III
 * Single Number III
 */
export function singleNumberIII(nums: number[]): number[] {
    let xor = 0;
    for (const num of nums) {
        xor ^= num;
    }
    
    // 找到最右边的1
    const rightmostBit = xor & (-xor);
    
    let num1 = 0;
    let num2 = 0;
    
    for (const num of nums) {
        if (num & rightmostBit) {
            num1 ^= num;
        } else {
            num2 ^= num;
        }
    }
    
    return [num1, num2];
}

/**
 * 191. 位1的个数
 * Number of 1 Bits
 */
export function hammingWeight(n: number): number {
    let count = 0;
    while (n !== 0) {
        count++;
        n &= n - 1; // 清除最右边的1
    }
    return count;
}

/**
 * 338. 比特位计数
 * Counting Bits
 */
export function countBits(n: number): number[] {
    const result = new Array(n + 1).fill(0);
    
    for (let i = 1; i <= n; i++) {
        result[i] = result[i >> 1] + (i & 1);
    }
    
    return result;
}

/**
 * 461. 汉明距离
 * Hamming Distance
 */
export function hammingDistance(x: number, y: number): number {
    let xor = x ^ y;
    let count = 0;
    
    while (xor !== 0) {
        count++;
        xor &= xor - 1;
    }
    
    return count;
}

/**
 * 477. 汉明距离总和
 * Total Hamming Distance
 */
export function totalHammingDistance(nums: number[]): number {
    let total = 0;
    const n = nums.length;
    
    for (let i = 0; i < 32; i++) {
        let ones = 0;
        for (const num of nums) {
            ones += (num >> i) & 1;
        }
        total += ones * (n - ones);
    }
    
    return total;
}

/**
 * 190. 颠倒二进制位
 * Reverse Bits
 */
export function reverseBits(n: number): number {
    let result = 0;
    for (let i = 0; i < 32; i++) {
        result = (result << 1) | (n & 1);
        n >>= 1;
    }
    return result >>> 0; // 无符号右移确保正数
}

/**
 * 231. 2的幂
 * Power of Two
 */
export function isPowerOfTwo(n: number): boolean {
    return n > 0 && (n & (n - 1)) === 0;
}

/**
 * 342. 4的幂
 * Power of Four
 */
export function isPowerOfFour(n: number): boolean {
    return n > 0 && (n & (n - 1)) === 0 && (n & 0x55555555) !== 0;
}

/**
 * 371. 两整数之和
 * Sum of Two Integers
 */
export function getSum(a: number, b: number): number {
    while (b !== 0) {
        const carry = (a & b) << 1;
        a = a ^ b;
        b = carry;
    }
    return a;
}

/**
 * 389. 找不同
 * Find the Difference
 */
export function findTheDifference(s: string, t: string): string {
    let xor = 0;
    
    for (const char of s) {
        xor ^= char.charCodeAt(0);
    }
    
    for (const char of t) {
        xor ^= char.charCodeAt(0);
    }
    
    return String.fromCharCode(xor);
}

/**
 * 405. 数字转换为十六进制数
 * Convert a Number to Hexadecimal
 */
export function toHex(num: number): string {
    if (num === 0) return "0";
    
    const hex = "0123456789abcdef";
    let result = "";
    
    // 处理负数
    if (num < 0) {
        num = 0x100000000 + num; // 转换为32位无符号数
    }
    
    while (num > 0) {
        result = hex[num & 15] + result;
        num >>>= 4;
    }
    
    return result;
}

/**
 * 421. 数组中两个数的最大异或值
 * Maximum XOR of Two Numbers in an Array
 */
export function findMaximumXOR(nums: number[]): number {
    let maxXor = 0;
    let mask = 0;
    
    for (let i = 30; i >= 0; i--) {
        mask |= (1 << i);
        const prefixes = new Set<number>();
        
        for (const num of nums) {
            prefixes.add(num & mask);
        }
        
        const candidate = maxXor | (1 << i);
        
        for (const prefix of prefixes) {
            if (prefixes.has(candidate ^ prefix)) {
                maxXor = candidate;
                break;
            }
        }
    }
    
    return maxXor;
}

/**
 * 476. 数字的补数
 * Number Complement
 */
export function findComplement(num: number): number {
    let mask = 1;
    let temp = num;
    
    while (temp > 0) {
        mask <<= 1;
        temp >>= 1;
    }
    
    return (mask - 1) ^ num;
}

/**
 * 693. 交替位二进制数
 * Binary Number with Alternating Bits
 */
export function hasAlternatingBits(n: number): boolean {
    const xor = n ^ (n >> 1);
    return (xor & (xor + 1)) === 0;
}

/**
 * 762. 二进制表示中质数个计算置位
 * Prime Number of Set Bits in Binary Representation
 */
export function countPrimeSetBits(left: number, right: number): number {
    const isPrime = (n: number): boolean => {
        if (n < 2) return false;
        if (n === 2) return true;
        if (n % 2 === 0) return false;
        
        for (let i = 3; i * i <= n; i += 2) {
            if (n % i === 0) return false;
        }
        return true;
    };
    
    let count = 0;
    
    for (let i = left; i <= right; i++) {
        const setBits = hammingWeight(i);
        if (isPrime(setBits)) {
            count++;
        }
    }
    
    return count;
}

/**
 * 868. 二进制间距
 * Binary Gap
 */
export function binaryGap(n: number): number {
    let maxGap = 0;
    let lastOne = -1;
    
    for (let i = 0; i < 32; i++) {
        if ((n >> i) & 1) {
            if (lastOne !== -1) {
                maxGap = Math.max(maxGap, i - lastOne);
            }
            lastOne = i;
        }
    }
    
    return maxGap;
}

/**
 * 1318. 或运算的最小翻转次数
 * Minimum Flips to Make a OR b Equal to c
 */
export function minFlips(a: number, b: number, c: number): number {
    let flips = 0;
    
    while (a > 0 || b > 0 || c > 0) {
        const bitA = a & 1;
        const bitB = b & 1;
        const bitC = c & 1;
        
        if (bitC === 0) {
            flips += bitA + bitB;
        } else {
            if (bitA === 0 && bitB === 0) {
                flips++;
            }
        }
        
        a >>= 1;
        b >>= 1;
        c >>= 1;
    }
    
    return flips;
}

File: algorithms/bit_manipulation.go
package algorithms

// 位操作相关算法实现
// Bit Manipulation related algorithms

// 136. 只出现一次的数字
// Single Number
// 在数组中找出只出现一次的数字（其他数字都出现两次）
// https://leetcode.cn/problems/single-number/description/
func SingleNumber(nums []int) int {
	result := 0
	for _, num := range nums {
		result ^= num
	}
	return result
}

// 137. 只出现一次的数字 II
// Single Number II
// 在数组中找出只出现一次的数字（其他数字都出现三次）
// https://leetcode.cn/problems/single-number-ii/description/
func SingleNumberII(nums []int) int {
	ones := 0
	twos := 0

	for _, num := range nums {
		ones = (ones ^ num) & ^twos
		twos = (twos ^ num) & ^ones
	}

	return ones
}

// 260. 只出现一次的数字 III
// Single Number III
// 在数组中找出两个只出现一次的数字（其他数字都出现两次）
// https://leetcode.cn/problems/single-number-iii/description/
func SingleNumberIII(nums []int) []int {
	xor := 0
	for _, num := range nums {
		xor ^= num
	}

	// 找到最右边的1
	rightmostBit := xor & (-xor)

	num1 := 0
	num2 := 0

	for _, num := range nums {
		if num&rightmostBit != 0 {
			num1 ^= num
		} else {
			num2 ^= num
		}
	}

	return []int{num1, num2}
}

// 191. 位1的个数
// Number of 1 Bits
// 计算一个无符号整数的二进制表示中1的个数
// https://leetcode.cn/problems/number-of-1-bits/description/
func HammingWeight(n uint32) int {
	count := 0
	for n != 0 {
		count++
		n &= n - 1 // 清除最右边的1
	}
	return count
}

// 338. 比特位计数
// Counting Bits
// 计算从0到n的每个数字的二进制表示中1的个数
// https://leetcode.cn/problems/counting-bits/description/
func CountBits(n int) []int {
	result := make([]int, n+1)

	for i := 1; i <= n; i++ {
		result[i] = result[i>>1] + (i & 1)
	}

	return result
}

// 461. 汉明距离
// Hamming Distance
// 计算两个整数二进制表示中不同位的个数
// https://leetcode.cn/problems/hamming-distance/description/
func HammingDistance(x, y int) int {
	xor := x ^ y
	count := 0

	for xor != 0 {
		count++
		xor &= xor - 1
	}

	return count
}

// 477. 汉明距离总和
// Total Hamming Distance
// 计算数组中所有数对之间的汉明距离总和
// https://leetcode.cn/problems/total-hamming-distance/description/
func TotalHammingDistance(nums []int) int {
	total := 0
	n := len(nums)

	for i := 0; i < 32; i++ {
		ones := 0
		for _, num := range nums {
			ones += (num >> i) & 1
		}
		total += ones * (n - ones)
	}

	return total
}

// 190. 颠倒二进制位
// Reverse Bits
// 颠倒给定的32位无符号整数的二进制位
// https://leetcode.cn/problems/reverse-bits/description/
func ReverseBits(n uint32) uint32 {
	var result uint32 = 0
	for i := 0; i < 32; i++ {
		result = (result << 1) | (n & 1)
		n >>= 1
	}
	return result
}

// 231. 2的幂
// Power of Two
// 判断一个整数是否是2的幂次方
// https://leetcode.cn/problems/power-of-two/description/
func IsPowerOfTwo(n int) bool {
	return n > 0 && (n&(n-1)) == 0
}

// 342. 4的幂
// Power of Four
// 判断一个整数是否是4的幂次方
// https://leetcode.cn/problems/power-of-four/description/
func IsPowerOfFour(n int) bool {
	return n > 0 && (n&(n-1)) == 0 && (n&0x55555555) != 0
}

// 371. 两整数之和
// Sum of Two Integers
// 不使用+和-运算符，计算两整数之和
// https://leetcode.cn/problems/sum-of-two-integers/description/
func GetSum(a, b int) int {
	for b != 0 {
		carry := (a & b) << 1
		a = a ^ b
		b = carry
	}
	return a
}

// 389. 找不同
// Find the Difference
// 在字符串t中找到比字符串s多出的那个字符
// https://leetcode.cn/problems/find-the-difference/description/
func FindTheDifference(s, t string) byte {
	xor := 0

	for _, char := range s {
		xor ^= int(char)
	}

	for _, char := range t {
		xor ^= int(char)
	}

	return byte(xor)
}

// 405. 数字转换为十六进制数
// Convert a Number to Hexadecimal
// 将给定的整数转换为十六进制字符串表示
// https://leetcode.cn/problems/convert-a-number-to-hexadecimal/description/
func ToHex(num int) string {
	if num == 0 {
		return "0"
	}

	hex := "0123456789abcdef"
	result := ""

	// 处理负数
	var n uint32
	if num < 0 {
		n = uint32(num)
	} else {
		n = uint32(num)
	}

	for n > 0 {
		result = string(hex[n&15]) + result
		n >>= 4
	}

	return result
}
